modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java

```diff
@@ -0,0 +1,322 @@
+    }
+
+    @Override
+    public long longValue(int col) {
+        return delegate.longValue(projection[col]);
+    }
+
+    @Override
+    public Long longValueBoxed(int col) {
+        return delegate.longValueBoxed(projection[col]);
+    }
+
+    @Override
+    public float floatValue(int col) {
+        return delegate.floatValue(projection[col]);
+    }
+
+    @Override
+    public Float floatValueBoxed(int col) {
+        return delegate.floatValueBoxed(projection[col]);
+    }
+
+    @Override
+    public double doubleValue(int col) {
+        return delegate.doubleValue(projection[col]);
+    }
+
+    @Override
+    public Double doubleValueBoxed(int col) {
+        return delegate.doubleValueBoxed(projection[col]);
+    }
+
+    @Override
+    public BigDecimal decimalValue(int col, int decimalScale) {
+        return delegate.decimalValue(projection[col], decimalScale);
+    }
+
+    @Override
+    public BigInteger numberValue(int col) {
+        return delegate.numberValue(projection[col]);
+    }
+
+    @Override
+    public String stringValue(int col) {
+        return delegate.stringValue(projection[col]);
+    }
+
+    @Override
+    public byte[] bytesValue(int col) {
+        return delegate.bytesValue(projection[col]);
+    }
+
+    @Override
+    public UUID uuidValue(int col) {
+        return delegate.uuidValue(projection[col]);
+    }
+
+    @Override
+    public BitSet bitmaskValue(int col) {
+        return delegate.bitmaskValue(projection[col]);
+    }
+
+    @Override
+    public LocalDate dateValue(int col) {
+        return delegate.dateValue(projection[col]);
+    }
+
+    @Override
+    public LocalTime timeValue(int col) {
+        return delegate.timeValue(projection[col]);
+    }
+
+    @Override
+    public LocalDateTime dateTimeValue(int col) {
+        return delegate.dateTimeValue(projection[col]);
+    }
+
+    @Override
+    public Instant timestampValue(int col) {
+        return delegate.timestampValue(projection[col]);
+    }
+
+    @Override
+    public ByteBuffer byteBuffer() {
+        normalizeIfNeeded();
+
+        return delegate.byteBuffer();
+    }
+
+    private void normalizeIfNeeded() {
+        if (normalized) {
+            return;
+        }
+
+        if (schema != null) {
+            normalizeSlow();
+        } else {
+            normalizeFast();
+        }
+    }
+
+    private void normalizeSlow() {
+        assert schema != null;
+
+        var builder = new BinaryTupleBuilder(projection.length);
+        var newProjection = new int[projection.length];
+
+        for (int i = 0; i < projection.length; i++) {
+            int col = projection[i];
+
+            newProjection[i] = i;
+
+            Element element = schema.element(col);
+
+            BinaryRowConverter.appendValue(builder, element, schema.value(delegate, col));
+        }
+
+        delegate = new BinaryTuple(projection.length, builder.build());
+        projection = newProjection;
+        normalized = true;
```

Comment:
seems `nomalized` flag should be moved to `normalizeIfNeeded`

Reply:
fixed